File: Sources/PalmierPro/Models/Effect.swift
import Foundation

/// One entry in a clip's ordered effect stack
struct Effect: Codable, Sendable, Equatable, Identifiable {
    var id: String = UUID().uuidString
    var type: String
    var enabled: Bool = true
    var params: [String: EffectParam] = [:]

    init(id: String = UUID().uuidString, type: String, enabled: Bool = true,
         params: [String: EffectParam] = [:]) {
        self.id = id
        self.type = type
        self.enabled = enabled
        self.params = params
    }

    private enum CodingKeys: String, CodingKey {
        case id, type, enabled, params
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            id: (try? c.decode(String.self, forKey: .id)) ?? UUID().uuidString,
            type: try c.decode(String.self, forKey: .type),
            enabled: (try? c.decode(Bool.self, forKey: .enabled)) ?? true,
            params: (try? c.decode([String: EffectParam].self, forKey: .params)) ?? [:]
        )
    }
}

/// A single effect parameter
struct EffectParam: Codable, Sendable, Equatable {
    var value: Double?
    var string: String?
    var track: KeyframeTrack<Double>?

    init(value: Double? = nil, string: String? = nil, track: KeyframeTrack<Double>? = nil) {
        self.value = value
        self.string = string
        self.track = track
    }

    /// Effective numeric value at a clip-relative frame offset.
    func resolved(at offset: Int, default defaultValue: Double) -> Double {
        if let track, track.isActive {
            return track.sample(at: offset, fallback: value ?? defaultValue)
        }
        return value ?? defaultValue
    }
}

extension Effect {
    /// Convenience for static numeric params.
    static func make(_ type: String, _ values: [String: Double] = [:]) -> Effect {
        Effect(type: type, params: values.mapValues { EffectParam(value: $0) })
    }
}

File: windows/src/PalmierPro.Core/Models/Effect.cs
using PalmierPro.Core.Serialization;

namespace PalmierPro.Core.Models;

/// <summary>One entry in a clip's ordered effect stack.</summary>
public sealed class Effect : IEquatable<Effect>
{
    public string Id { get; set; } = Uuid.NewString();
    public required string Type { get; set; }
    public bool Enabled { get; set; } = true;
    public Dictionary<string, EffectParam> Params { get; set; } = [];

    public static Effect Make(string type, IReadOnlyDictionary<string, double>? values = null)
    {
        var effect = new Effect { Type = type };
        if (values is not null)
        {
            foreach (var (key, value) in values)
            {
                effect.Params[key] = new EffectParam { Value = value };
            }
        }
        return effect;
    }

    public bool Equals(Effect? other)
        => other is not null
            && Id == other.Id
            && Type == other.Type
            && Enabled == other.Enabled
            && Params.Count == other.Params.Count
            && Params.All(kv => other.Params.TryGetValue(kv.Key, out var v) && kv.Value.Equals(v));

    public override bool Equals(object? obj) => Equals(obj as Effect);
    public override int GetHashCode() => HashCode.Combine(Id, Type, Enabled);
}

/// <summary>A single effect parameter.</summary>
public sealed class EffectParam : IEquatable<EffectParam>
{
    public double? Value { get; set; }
    public string? String { get; set; }
    public KeyframeTrack<double>? Track { get; set; }

    /// <summary>Effective numeric value at a clip-relative frame offset.</summary>
    public double Resolved(int offset, double defaultValue)
    {
        if (Track is { IsActive: true })
        {
            return Track.Sample(offset, Value ?? defaultValue);
        }
        return Value ?? defaultValue;
    }

    public bool Equals(EffectParam? other)
        => other is not null
            && Value == other.Value
            && String == other.String
            && Equals(Track, other.Track);

    public override bool Equals(object? obj) => Equals(obj as EffectParam);
    public override int GetHashCode() => HashCode.Combine(Value, String);
}
